From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrii Sultanov <andriy.sultanov@vates.tech>
Date: Mon, 18 May 2026 13:13:33 +0000
Subject: [PATCH] storage: Preserve VDI tags on SMAPIv1 migrate

SMAPI.VDI.create ignores the tags from vdi_info, so set them after the call

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>
(cherry picked from commit 76c7e267da596ba4a527457c318ce81f19b38ef4)
---
 ocaml/xapi-storage-script/main.ml     | 8 ++++++++
 ocaml/xapi/storage_smapiv1_migrate.ml | 5 +++++
 2 files changed, 13 insertions(+)

diff --git a/ocaml/xapi-storage-script/main.ml b/ocaml/xapi-storage-script/main.ml
index 10ca7689f..1b72d5a8f 100644
--- a/ocaml/xapi-storage-script/main.ml
+++ b/ocaml/xapi-storage-script/main.ml
@@ -1479,6 +1479,14 @@ module VDIImpl (M : META) = struct
     set ~dbg ~sr ~vdi:response.Xapi_storage.Control.key ~key:_vdi_type_key
       ~value:vdi_info.ty
     >>>= fun () ->
+    let rec f = function
+      | key :: x ->
+          let* _ = set ~dbg ~sr ~vdi ~key:(_vdi_tags_key ^ key) ~value:key in
+          f x
+      | [] ->
+          return ()
+    in
+    f vdi_info.tags >>>= fun () ->
     let response =
       {
         (vdi_of_volume response) with
diff --git a/ocaml/xapi/storage_smapiv1_migrate.ml b/ocaml/xapi/storage_smapiv1_migrate.ml
index b201813d8..b8d43ee79 100644
--- a/ocaml/xapi/storage_smapiv1_migrate.ml
+++ b/ocaml/xapi/storage_smapiv1_migrate.ml
@@ -650,6 +650,11 @@ module MIRROR : SMAPIv2_MIRROR = struct
             add_to_sm_config vdi_info "image-format" fmt
       in
       let leaf = SMAPI.VDI.create dbg sr vdi_info in
+
+      List.iter
+        (fun tag -> Local.VDI.add_tags dbg sr leaf.vdi tag)
+        vdi_info.tags ;
+
       D.info "Created leaf VDI for mirror receive: %s" (string_of_vdi_info leaf) ;
       on_fail := (fun () -> SMAPI.VDI.destroy dbg sr leaf.vdi) :: !on_fail ;
       (* dummy VDI is created so that the leaf VDI becomes a differencing disk,
